staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go

```diff
@@ -132,6 +132,9 @@ func NestedStringSlice(obj map[string]interface{}, fields ...string) ([]string,
 	if !found || err != nil {
 		return nil, found, err
 	}
+	if val == nil {
+		return nil, true, nil
```

Comment:
I'm not sure we want to change this for all callers... a more targeted fix for kubectl wait would be easier to reason about

```diff
diff --git a/staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go b/staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go
index 83ca3fc910e..4f20c612086 100644
--- a/staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go
+++ b/staging/src/k8s.io/kubectl/pkg/cmd/wait/wait.go
@@ -488,15 +488,25 @@ func (w ConditionalWait) IsConditionMet(info *resource.Info, o *WaitOptions) (ru
 }
 
 func (w ConditionalWait) checkCondition(obj *unstructured.Unstructured) (bool, error) {
-	conditions, found, err := unstructured.NestedSlice(obj.Object, "status", "conditions")
+	val, found, err := unstructured.NestedFieldNoCopy(obj.Object, "status", "conditions")
 	if err != nil {
 		return false, err
 	}
 	if !found {
 		return false, nil
 	}
-	for _, conditionUncast := range conditions {
-		condition := conditionUncast.(map[string]interface{})
+	if val == nil {
+		return false, nil
+	}
+	conditions, ok := val.([]interface{})
+	if !ok {
+		return false, fmt.Errorf(".status.conditions is of the type %T, expected []interface{}", val)
+	}
+	for i, conditionUncast := range conditions {
+		condition, ok := conditionUncast.(map[string]interface{})
+		if !ok {
+			return false, fmt.Errorf(".status.conditions[%d] is of the type %T, expected map[string]interface{}", i, conditionUncast)
+		}
 		name, found, err := unstructured.NestedString(condition, "type")
 		if !found || err != nil || !strings.EqualFold(name, w.conditionName) {
 			continue
```

Reply:
I considered that, but `nil` is documented in the Go spec as the empty slice value. As it is, this helper is introducing a type restriction that the language itself does not. With my change, the helper also returns the empty slice value, which the caller can use as an empty slice without worrying about it. I am wary of forcing the caller to make even more type assertions only to preserve this non-standard behaviour.

A fix for kubectl is more targeted for the particular context in which I encountered it, but it keeps the complexity that this part of apimachinery has a narrower slice type than Go. I find this harder to reason about, not easier. If we do consider this [breaking the workflow](https://xkcd.com/1172/), when can this difference be resolved?